diff -pur dosfstools-2.11/dosfsck/boot.c dosfstools-2.11.new/dosfsck/boot.c
--- dosfstools-2.11/dosfsck/boot.c	2005-03-12 16:16:22.000000000 +0200
+++ dosfstools-2.11.new/dosfsck/boot.c	2007-06-15 12:59:21.897962824 +0300
@@ -34,13 +34,13 @@ static struct {
     { 0xff, "5.25\" 320k floppy 2s/40tr/8sec" },
 };
 
-#if defined __alpha || defined __ia64__ || defined __s390x__ || defined __x86_64__ || defined __ppc64__
+#if defined __alpha || defined __ia64__ || defined __s390x__ || defined __x86_64__ || defined __ppc64__ || __arm__
 /* Unaligned fields must first be copied byte-wise */
 #define GET_UNALIGNED_W(f)			\
     ({						\
 	unsigned short __v;			\
 	memcpy( &__v, &f, sizeof(__v) );	\
-	CF_LE_W( *(unsigned short *)&f );	\
+	CF_LE_W( *(unsigned short *)&__v );	\
     })
 #else
 #define GET_UNALIGNED_W(f) CF_LE_W( *(unsigned short *)&f )
Only in dosfstools-2.11.new/dosfsck: boot.c~
