diff -pur dosfstools-2.11/dosfsck/fat.c dosfstools-2.11.new/dosfsck/fat.c
--- dosfstools-2.11/dosfsck/fat.c	2003-06-14 23:03:20.000000000 +0300
+++ dosfstools-2.11.new/dosfsck/fat.c	2007-10-04 10:05:32.673762744 +0300
@@ -53,6 +53,7 @@ void read_fat(DOS_FS *fs)
     unsigned long i;
     void *first,*second,*use;
     int first_ok,second_ok;
+    int msg_to_print = 100, total_msg = 0, msg_printed = 0;
 
     eff_size = ((fs->clusters+2)*fs->fat_bits+7)/8;
     first = alloc(eff_size);
@@ -102,10 +103,18 @@ void read_fat(DOS_FS *fs)
     for (i = 2; i < fs->clusters+2; i++)
 	if (fs->fat[i].value >= fs->clusters+2 &&
 	    (fs->fat[i].value < FAT_MIN_BAD(fs))) {
+            if (msg_to_print >= msg_printed) {
 	    printf("Cluster %ld out of range (%ld > %ld). Setting to EOF.\n",
 		   i-2,fs->fat[i].value,fs->clusters+2-1);
+            ++msg_printed;
+            }
 	    set_fat(fs,i,-1);
+            ++total_msg;
 	}
+    if (total_msg - msg_printed > 0) {
+        printf("Skipped %d 'Cluster N out of range...' messages.\n",
+               total_msg - msg_printed);
+    }
     free(first);
     if (second)
 	free(second);
