--- dosfstools-2.11/dosfsck/check.c.orig	2008-02-07 16:11:26.000000000 +0200
+++ dosfstools-2.11/dosfsck/check.c	2008-02-07 16:13:22.000000000 +0200
@@ -315,8 +315,8 @@
     first = file->parent ? file->parent->first : root;
     number = 0;
     while (1) {
-	sprintf(file->dir_ent.name, "FSCK%04d", number / 1000);
-	sprintf(file->dir_ent.name, "%03d", number % 1000);
+	sprintf(file->dir_ent.name, "FSCK%04ld", number / 1000);
+	sprintf(file->dir_ent.name, "%03ld", number % 1000);
 	for (walk = first; walk; walk = walk->next)
 	    if (walk != file && !strncmp(walk->dir_ent.name,file->dir_ent.
 	      name,MSDOS_NAME)) break;
@@ -535,7 +535,7 @@
     if (!(file->dir_ent.attr & ATTR_DIR) && CF_LE_L(file->dir_ent.size) >
       (unsigned long long)clusters*fs->cluster_size) {
 	printf("%s\n  File size is %u bytes, cluster chain length is %llu bytes."
-	  "\n  Truncating file to %lu bytes.\n",path_name(file),CF_LE_L(file->
+	  "\n  Truncating file to %llu bytes.\n",path_name(file),CF_LE_L(file->
 	  dir_ent.size),(unsigned long long)clusters*fs->cluster_size,(unsigned long long)clusters*fs->cluster_size);
 	MODIFY(file,size,CT_LE_L((unsigned long long)clusters*fs->cluster_size));
     }
