--- portmap_5beta/Makefile	2009-02-11 11:41:33.000000000 +0100
+++ portmap_new/Makefile	2009-02-11 12:08:47.000000000 +0100
@@ -88,7 +88,7 @@
 LIBS	= -lutil
 NSARCHS	= 
 O	= -Wall -O2 -pipe
-
+COMPSTATIC = -static
 # NEXTSTEP is a little different. The following seems to work with NS 3.2
 #
 # SETPGRP	=-DUSE_SETPGRP00
@@ -119,6 +119,13 @@
 portmap: $(OBJECTS)
 	$(CC) $(CFLAGS) -o $@ $(OBJECTS) $(WRAP_LIB) $(LIBS)
 
+portmap-mkdir-static: 
+	/bin/mkdir ./portmap-bin 
+
+portmap-static: $(OBJECTS)
+	$(CC) $(CFLAGS) $(COMPSTATIC) -o portmap-bin/portmap-static $(OBJECTS) $(WRAP_LIB) $(LIBS)
+	
+
 pmap_dump: pmap_dump.c
 	$(CC) $(CFLAGS) -o $@ $? $(LIBS)
 
@@ -135,7 +142,7 @@
 	lint $(COPT) $(OBJECTS:%.o=%.c)
 
 clean:
-	rm -f *.o portmap pmap_dump pmap_set from_local get_myaddress \
+	rm -f *.o portmap portmap-bin/portmap-static pmap_dump pmap_set from_local get_myaddress \
 	    get_myaddress.so core
 
 tidy:	clean
