diff -Nur uClibc_bak/libc/sysdeps/linux/common/bits/termios.h uClibc-0.9.29/libc/sysdeps/linux/common/bits/termios.h
--- uClibc_bak/libc/sysdeps/linux/common/bits/termios.h	2009-01-27 11:45:31.000000000 +0100
+++ uClibc-0.9.29/libc/sysdeps/linux/common/bits/termios.h	2009-01-27 11:47:31.000000000 +0100
@@ -156,7 +156,7 @@
 #endif
 #define  B57600   0010001
 #define  B115200  0010002
-#if 0 /* limited on uClibc, keep in sync w/ cfsetspeed.c */
+#if 1 /* limited on uClibc, keep in sync w/ cfsetspeed.c */
 #define  B230400  0010003
 #define  B460800  0010004
 #define  B500000  0010005
