diff -ruN -X /home/marc/dontdiff libupnp-1.6.6.orig/build/inc/config.h libupnp-1.6.6/build/inc/config.h
--- libupnp-1.6.6.orig/build/inc/config.h	2008-02-10 02:55:38.000000000 +0100
+++ libupnp-1.6.6/build/inc/config.h	2009-12-08 14:45:05.000000000 +0100
@@ -113,7 +113,7 @@
  * This can be adjusted dynamically with {\tt UpnpSetMaxContentLength}.
  */
 //@{
-#define DEFAULT_SOAP_CONTENT_LENGTH 16000
+#define DEFAULT_SOAP_CONTENT_LENGTH 2000000
 //@}
 
 /** @name NUM_SSDP_COPY
