diff -uwbr libmad-0.15.1b_org/configure.ac libmad-0.15.1b/configure.ac
--- libmad-0.15.1b_org/configure.ac	2004-01-23 10:41:32.000000000 +0100
+++ libmad-0.15.1b/configure.ac	2008-11-17 14:13:07.000000000 +0100
@@ -140,7 +140,7 @@
     case "$optimize" in
 	-O|"-O "*)
 	    optimize="-O"
-	    optimize="$optimize -fforce-mem"
+	    optimize="$optimize"
 	    optimize="$optimize -fforce-addr"
 	    : #x optimize="$optimize -finline-functions"
 	    : #- optimize="$optimize -fstrength-reduce"
