diff -urN fuse-2.5.3/util/Makefile.in fuse-2.5.3_mod/util/Makefile.in
--- fuse-2.5.3/util/Makefile.in	2006-04-10 10:47:13.000000000 +0200
+++ fuse-2.5.3_mod/util/Makefile.in	2006-05-04 12:12:03.000000000 +0200
@@ -453,11 +453,6 @@
 install-exec-hook:
 	-chown root $(DESTDIR)$(bindir)/fusermount
 	-chmod u+s $(DESTDIR)$(bindir)/fusermount
-	@if test ! -e $(DESTDIR)/dev/fuse; then \
-		$(mkdir_p) $(DESTDIR)/dev; \
-		echo "mknod $(DESTDIR)/dev/fuse -m 0666 c 10 229"; \
-		mknod $(DESTDIR)/dev/fuse -m 0666 c 10 229; \
-	fi
 
 install-exec-local:
 	$(mkdir_p) $(DESTDIR)$(MOUNT_FUSE_PATH)
