--- fuse-2.6.0.old/util/Makefile.am	2010-05-31 16:51:29.000000000 +0200
+++ fuse-2.6.0/util/Makefile.am	2010-05-31 16:51:58.000000000 +0200
@@ -12,7 +12,7 @@
 install-exec-hook:
 	-chown root $(DESTDIR)$(bindir)/fusermount
 	-chmod u+s $(DESTDIR)$(bindir)/fusermount
-	@if test ! -e $(DESTDIR)/dev/fuse; then \
+#	@if test ! -e $(DESTDIR)/dev/fuse; then \
 		$(mkdir_p) $(DESTDIR)/dev; \
 		echo "mknod $(DESTDIR)/dev/fuse -m 0666 c 10 229"; \
 		mknod $(DESTDIR)/dev/fuse -m 0666 c 10 229; \
@@ -27,9 +27,9 @@
 install-exec-local:
 	$(mkdir_p) $(DESTDIR)$(MOUNT_FUSE_PATH)
 	$(INSTALL_PROGRAM) $(srcdir)/mount.fuse $(DESTDIR)$(MOUNT_FUSE_PATH)/mount.fuse
-	$(mkdir_p) $(DESTDIR)$(INIT_D_PATH)
-	$(INSTALL_PROGRAM) $(srcdir)/init_script $(DESTDIR)$(INIT_D_PATH)/fuse
-	@if test -x /usr/sbin/update-rc.d; then \
+#	$(mkdir_p) $(DESTDIR)$(INIT_D_PATH)
+#	$(INSTALL_PROGRAM) $(srcdir)/init_script $(DESTDIR)$(INIT_D_PATH)/fuse
+#	@if test -x /usr/sbin/update-rc.d; then \
 		echo "/usr/sbin/update-rc.d fuse start 34 S . start 41 0 6 ."; \
 		/usr/sbin/update-rc.d fuse start 34 S . start 41 0 6 .; \
 	fi
@@ -41,8 +41,8 @@
 uninstall-local:
 	rm -f $(DESTDIR)$(MOUNT_FUSE_PATH)/mount.fuse
 	rm -f $(DESTDIR)$(UDEV_RULES_PATH)/99-fuse.rules
-	rm -f $(DESTDIR)$(INIT_D_PATH)/fuse
-	@if test -x /usr/sbin/update-rc.d; then \
+#	rm -f $(DESTDIR)$(INIT_D_PATH)/fuse
+#	@if test -x /usr/sbin/update-rc.d; then \
 		echo "/usr/sbin/update-rc.d fuse remove"; \
 		/usr/sbin/update-rc.d fuse remove; \
 	fi
--- fuse-2.6.0.old/util/Makefile.in	2010-05-31 16:51:29.000000000 +0200
+++ fuse-2.6.0/util/Makefile.in	2010-05-31 16:51:58.000000000 +0200
@@ -493,9 +493,9 @@
 install-exec-local:
 	$(mkdir_p) $(DESTDIR)$(MOUNT_FUSE_PATH)
 	$(INSTALL_PROGRAM) $(srcdir)/mount.fuse $(DESTDIR)$(MOUNT_FUSE_PATH)/mount.fuse
-	$(mkdir_p) $(DESTDIR)$(INIT_D_PATH)
-	$(INSTALL_PROGRAM) $(srcdir)/init_script $(DESTDIR)$(INIT_D_PATH)/fuse
-	@if test -x /usr/sbin/update-rc.d; then \
+#	$(mkdir_p) $(DESTDIR)$(INIT_D_PATH)
+#	$(INSTALL_PROGRAM) $(srcdir)/init_script $(DESTDIR)$(INIT_D_PATH)/fuse
+#	@if test -x /usr/sbin/update-rc.d; then \
 		echo "/usr/sbin/update-rc.d fuse start 34 S . start 41 0 6 ."; \
 		/usr/sbin/update-rc.d fuse start 34 S . start 41 0 6 .; \
 	fi
@@ -507,8 +507,8 @@
 uninstall-local:
 	rm -f $(DESTDIR)$(MOUNT_FUSE_PATH)/mount.fuse
 	rm -f $(DESTDIR)$(UDEV_RULES_PATH)/99-fuse.rules
-	rm -f $(DESTDIR)$(INIT_D_PATH)/fuse
-	@if test -x /usr/sbin/update-rc.d; then \
+#	rm -f $(DESTDIR)$(INIT_D_PATH)/fuse
+#	@if test -x /usr/sbin/update-rc.d; then \
 		echo "/usr/sbin/update-rc.d fuse remove"; \
 		/usr/sbin/update-rc.d fuse remove; \
 	fi
